### ISSUE TITLE
getdents() / lseek() behaviour is not transparent on mountpoint

### ISSUE THREAD

**Issue by dtakken** [mpartel/bindfs#41]
Hello,

Recently, some software that I am running on a development box started misbehaving. The setup uses bindfs to configure the correct uid / gid on the files for use inside a Docker container. As it turned out, the sofware reading from the bindfs mount reported any directory inside the mount point to be empty, while in fact they were not. Running the same software without the bind mount does not show this problem.

Running the software with strace reveals that the software is running a specific sequence of file system operations:

```
openat(AT_FDCWD, "/data/", O_RDONLY|O_NONBLOCK|O_DIRECTORY|O_CLOEXEC) = 3
getdents(3, /* 30 entries */, 32768)    = 1080
lseek(3, 0, SEEK_SET)                   = 0
getdents(3, /* 0 entries */, 32768)     = 0
```

So, it reads the directory entry, which yields 1080 bytes of data. Then, it rewinds the file pointer and reads the directory entry again. The second read yields no data, which apparently makes the software believe that the directory is empty. Without the bind mount, the problem does not occur and strace shows:

```
openat(AT_FDCWD, "/data/", O_RDONLY|O_NONBLOCK|O_DIRECTORY|O_CLOEXEC) = 3
getdents(3, /* 30 entries */, 32768)    = 1080
lseek(3, 0, SEEK_SET)                   = 0
getdents(3, /* 30 entries */, 32768)    = 1080
```

I wrote a test which reproduces the problem, at least on one system. Below, a small C program which executes the operations shown above, printing the two byte counts on the terminal:

```
#define _GNU_SOURCE
#include <dirent.h>     /* Defines DT_* constants */
#include <fcntl.h>
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <sys/syscall.h>

#define handle_error(msg) \
        do { perror(msg); exit(EXIT_FAILURE); } while (0)

struct linux_dirent {
    long           d_ino;
    off_t          d_off;
    unsigned short d_reclen;
    char           d_name[];
};

#define BUF_SIZE 1024

int
main(int argc, char *argv[])
{
    int fd, nread;
    char buf[BUF_SIZE];
    struct linux_dirent *d;
    int bpos;
    char d_type;

   fd = open(argc > 1 ? argv[1] : ".", O_RDONLY | O_DIRECTORY);
    if (fd == -1)
        handle_error("open");

   // Read from directory entry, print the
   // number of bytes read
   nread = syscall(SYS_getdents, fd, buf, BUF_SIZE);
   printf("%d\n", nread);

   // Rewind file pointer
   lseek(fd, 0, SEEK_SET);

   // Read from directory entry again, print the
   // number of bytes read
   nread = syscall(SYS_getdents, fd, buf, BUF_SIZE);
   printf("%d\n", nread);

   exit(EXIT_SUCCESS);
}
```

Below, a shell script that creates a bind mount and runs the above code on both the original directory and its bind mount:

```
#!/bin/sh
# Create two directories
mkdir data datamount
# Bind mount data/ on datamount/
bindfs data datamount/
# Compile source
gcc test.c -o test
# Run test executable on both directories, output
# should be equal.
echo "Read from original directory:"
./test data/
echo "Read from bind mount:"
./test datamount/
# Unmount
fusermount -u datamount/
# Remove directories
rmdir data datamount
```

On Ubuntu Trusty, kernel 3.13.0-98, fuse 2.9.2, bindfs 1.12.3-1, I get the result that, I guess, shows the expected result, the bind mount is fully transparent:

```
Read from original directory:
48
48
Read from bind mount:
48
48
```

On Suse Tumbleweed, kernel 4.7.5-1, fuse 2.9.5-1.5, bindfs 1.13.2-1.1, I get something different:

```
Read from original directory:
48
48
Read from bind mount:
48
0
```

I have no idea if this is problem is caused by bindfs, fuse, the Linux kernel or something else. Or perhaps it is not a bug at all. Could someone please try to enlighten me here?

Thanks!

**Comment by mpartel** [comment]
Thanks for the thorough report!

I believe this bug was introduced in 1.13.1 and it should now be fixed in 1.13.4.

**Comment by dtakken** [comment]
Thorough reports generally make quick fixes. :) And indeed, you managed to fix it in no time, thanks a lot!
